### PR TITLE
TE 3429 Create a Copy module spryk

### PR DESCRIPTION
**Pull request by stereomon**
- Developer(s): @stereomon

- Ticket: https://spryker.atlassian.net/browse/TE-3428

- Academy PR: ACADEMY_URL_HERE

- Release Group: URL_HERE

#### Please confirm

- [ ] No new OS components - or they have been approved by legal department

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Spryk                 | minor (currently 0.x)  |                       |

#### Release Notes

Added new CopyModule spryk to be able to copy a module to another location. Fixed overriding of argument values from CLI. Updated spryks which made use of `method` argument and changed the argument name to be more specific e.g. `facadeMethod`, `modelMethod` etc.

-----------------------------------------

##### Change log

Improvements

- Added new CopyModule spryk to be able to copy a module to another location. 
- Fixed overriding of argument values from CLI. 
- Updated spryks which made use of `method` argument and changed the argument name to be more specific e.g. `facadeMethod`, `modelMethod` etc.

### CODE REVIEW

**Comment by dereuromark** [comment]
Is copying a real use case? and only for core?

**Comment by stereomon** [comment]
> Is copying a real use case? and only for core?

Yes, I do that often and I think others also copied modules already manually. It's only for core for now as it would require much more effort to be usable in projects.

This command is also very very helpful when you apparently created a new module with a wrong name or a typo.

**Comment by dereuromark** [comment]
Do we need to major here? Can't we release this as minor changes (patch release)? Otherwise GUI module also needs a major.

File: src/SprykerSdk/Spryk/Model/Spryk/Definition/Argument/Resolver/ArgumentResolver.php [hunk 1]
```diff
@@ -156,7 +162,7 @@ protected function canInherit(string $argumentName, array $argumentDefinition):
      */
     protected function getKnownValueForArgument(string $argumentName, array $argumentDefinition)
     {
-        if (isset($argumentDefinition['value'])) {
+        if (isset($argumentDefinition['value']) && (!OptionsContainer::hasOption($argumentName) || empty(OptionsContainer::getOption($argumentName)))) {
```
Comment (dereuromark) [inline src/SprykerSdk/Spryk/Model/Spryk/Definition/Argument/Resolver/ArgumentResolver.php]: We dont use empty() on methods. ! suffices here.

File: src/SprykerSdk/Spryk/Model/Spryk/Definition/Argument/Resolver/ArgumentResolver.php [hunk 2]
```diff
@@ -193,13 +199,24 @@ protected function getDefaultValue(string $argumentName, ?array $argumentDefinit
      * @param string $argument
      * @param string $sprykName
      * @param string|int|null $default
+     * @param bool|null $allowEmpty
```
Comment (dereuromark) [inline src/SprykerSdk/Spryk/Model/Spryk/Definition/Argument/Resolver/ArgumentResolver.php]: why do we need null if the default could be false?

File: tests/SprykerSdkTest/Spryk/Integration/Client/AddClientConfigMethodTest.php [hunk 3]
```diff
@@ -44,16 +44,16 @@ public function testAddsClientConfigMethod(): void
...
-            '--module' => 'FooBar',
-            '--method' => 'getConfigValue',
-            '--input' => 'string $foo',
-            '--output' => 'string',
-            '--mode' => 'project',
-        ]);
-
-        $this->tester->assertClassHasMethod(ClassName::PROJECT_CLIENT_CONFIG, 'getConfigValue');
-    }
+//    public function testAddsClientConfigMethodOnProjectLayer(): void
```
Comment (gechetspr) [inline tests/SprykerSdkTest/Spryk/Integration/Client/AddClientConfigMethodTest.php]: Do we need it?

Comment (stereomon) [reply tests/SprykerSdkTest/Spryk/Integration/Client/AddClientConfigMethodTest.php]: Probably yes, I reactivated the test

File: src/SprykerSdk/Spryk/Model/Spryk/Definition/Argument/Resolver/ArgumentResolver.php (hunk 2 above)
Comment (dereuromark) [reply src/SprykerSdk/Spryk/Model/Spryk/Definition/Argument/Resolver/ArgumentResolver.php]: the null part should still be removed IMO

**Pull request merged**